IntegrationTests/UITests/UITest.cs

```diff
@@ -1,58 +1,181 @@
+                test.Join();
+                LoggingService.Log($"{MethodBase.GetCurrentMethod().Name} test joined");
+            }
+            finally
+            {
+                Application.Idle -= HandleApplicationIdle;
+                LoggingService.Log($"{MethodBase.GetCurrentMethod().Name} unsubscribed");
+
+                // Safety net for the case the task "test" threw very early and didn't close the form.
+                var form = Application.OpenForms.OfType<T>().SingleOrDefault();
+                if (form != null)
+                {
+                    string formState = form.Visible ? "open" : "exists";
+                    LoggingService.Log($"{MethodBase.GetCurrentMethod().Name} {typeof(T).FullName} still {formState}");
+
+                    CloseOrDispose(form, $"{MethodBase.GetCurrentMethod().Name} {typeof(T).FullName}");
+                }
+
+                Assert.IsTrue(Application.OpenForms.Count == 0, $"{Application.OpenForms.Count} open form(s)");
+            }
 
-            await idleCompletionSource.Task;
-            Application.Idle -= HandleApplicationIdle;
+            return;
 
             void HandleApplicationIdle(object sender, EventArgs e)
             {
+                LoggingService.Log($"{MethodBase.GetCurrentMethod().Name}");
                 idleCompletionSource.TrySetResult(default);
             }
         }
 
-        public static void RunForm<T>(
-            Action showDialog,
-            Func<T, Task> runAsync)
-            where T : Form
+        private static void CloseOrDispose(Form form, string description)
         {
-            // Avoid using ThreadHelper.JoinableTaskFactory for the outermost operation because we don't want the task
-            // tracked by its collection. Otherwise, test code would not be able to wait for pending operations to
-            // complete.
-            var test = ThreadHelper.JoinableTaskContext.Factory.RunAsync(async () =>
+            try
             {
-                await ThreadHelper.JoinableTaskFactory.SwitchToMainThreadAsync();
-                await WaitForIdleAsync();
-                var dialog = Application.OpenForms.OfType<T>().Single();
+                if (form == null)
+                {
+                    LoggingService.Log($"{description} form is null");
+                    return;
+                }
+
+                var actionDescription = description + " form ";
                 try
                 {
-                    await runAsync(dialog);
+                    if (form.Visible)
+                    {
+                        actionDescription += "close";
+                        form.Close(); // also calls form.Dispose()
+                    }
+                    else
+                    {
+                        actionDescription += "dispose";
+                        form.Dispose();
+                    }
```

Comment:
❔ Why not just call `Close`?

Reply:
Because I do not and should not know the implementation of `Close`. The documentation just states "Closes the form." That's why the obvious assumption is "nothing will happen if the form has already been closed or never been shown".

Reply:
On review, would should do both.

```
form.Close();
form.Dispose();
```

Reply:
The easiest way to implement this is change the original `var dialog` to be a using statement:

```patch
-var dialog = Application.OpenForms.OfType<T>().Single();
+using var dialog = Application.OpenForms.OfType<T>().Single();
```

Reply:
💭 As far as I can tell, this entire pull request can be simplified to this one line change.

Reply:
`Dispose` does not call neither `Close` nor `OnFormClose`, etc.

Reply:
The test already calls `Close`. You were correct that `Dispose` needs to be called as well. Placing the `var dialog` in a using declaration statement ensures that the call to `Dispose` will occur after the call to `Close`.